sources/models/src/modeled_types/credentialproviders.rs

```diff
@@ -0,0 +1,132 @@
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+// Just need serde's Error in scope to get its trait methods
+use super::error;
+use serde::de::Error as _;
+use snafu::ensure;
+use std::borrow::Borrow;
+use std::convert::TryFrom;
+use std::fmt;
+use std::ops::Deref;
+
+use crate::SingleLineString;
+
+/// CacheDurationValue is a string that represents the duration the plugin will
+/// cache credentials in-memory if a cache duration is not provided in the
+/// plugin response.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct CacheDurationValue {
+    inner: String,
+}
+
+lazy_static! {
+    // A duration string is a sequence of decimal numbers, each with a
+    // unit suffix, such as "300ms" or "2h45m".
+    // Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+    // Based on restraints from Go's `ParseDuration` funciton.
+    pub(crate) static ref DURATION_FORMAT: Regex = Regex::new(r"^([0-9]*[a-z]+)+$").unwrap();
+}
+
+impl TryFrom<&str> for CacheDurationValue {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, error::Error> {
+        ensure!(
+            DURATION_FORMAT.is_match(input),
+            error::InvalidCacheDurationValueSnafu { input }
+        );
+        Ok(CacheDurationValue {
+            inner: input.to_string(),
+        })
+    }
+}
+
+string_impls_for!(CacheDurationValue, "CacheDurationValue");
+
+#[cfg(test)]
+mod test_cacheduration_value {
+    use super::CacheDurationValue;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn valid_cacheduration_value() {
+        for ok in &["12h", "120m", "2h15m"] {
+            CacheDurationValue::try_from(*ok).unwrap();
+        }
+    }
+
+    #[test]
+    fn invalid_cacheduration_value() {
+        for err in &["12 days", "💝", "Ï", ""] {
+            CacheDurationValue::try_from(*err).unwrap_err();
+        }
+    }
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// CredentialPluginName is the name of the plugin which MUST match the
+/// name of the executable binary that exists in the Kubelet
+/// `--image-credential-provider-bin-dir` directory.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct CredentialPluginName {
+    inner: String,
+}
+
+lazy_static! {
+    // List of valid, supported credential provider plugins. Update this list
+    // if/when we ever add or remove credential provider plugin binaries.
+    pub(crate) static ref CREDENTIALPROVIDER_PLUGINS: [&'static str; 1] = ["ecr-credential-provider"];
```

Comment:
Instead of a hard-coded list, what do you think about validating at runtime that the provided plugin name exists in the right place on the filesystem?

That would eliminate the need to touch this part of the code after packaging a plugin and including it in the image.

Reply:
We could just remove this then. Kubelet will check that the named binary exists in the `--image-credential-provider-bin-dir` directory. Would that be preferable?

Reply:
I've changed this input type to be similar to `SingleLineString` with the addition of checking for any path traversal characters. So we can at least make sure someone isn't trying to access something they shouldn't be, but we can leave the actual validation of the plugin binary to kubelet.

With the previous way, if the name was not correct it would actual stall the boot of an instance and it wasn't very apparent to the user what was happening and what they needed to do to resolve it. With simplifying our checks here, the node will now boot with an invalid plugin name, but kubelet will error and log a message that it can't find the plugin. Then the instance will at least be running and accessible, and the user has a better chance to view the logs and see what the root of the issue is.